### PR TITLE
Allow to build packages for nix

### PR DESCRIPTION
**Pull request by romanofski**
This adds a few overwrites needed for packages which are not in the known good
set for GHC 8.2.2. Also, we pin Cabal with which we won't be able to build
purebred.

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+let
+  config = {
+    packageOverrides = super: let self = super.pkgs; in
+    {
+      haskell = super.haskell // {
+        packages = super.haskell.packages // {
+          "${compiler}" = super.haskell.packages."${compiler}".override {
+            overrides = self: super: {
+              # Build with latest known stable version.
+              purebred = with pkgs.haskell.lib; dontHaddock (self.callPackage ./.nix-helpers/purebred.nix { });
```
Comment (frasertweedale) [inline default.nix]: Why not build the haddock?  Is there a technical reason why we disable it?

File: default.nix [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+              purebred-email = self.callPackage ./.nix-helpers/purebred-email.nix { };
+              # Version pins since those versions were bleeding edge and not on
+              # hackage/stackage or Nix yet. Remove these once builds without
+              # these pass.
+              notmuch = self.callPackage ./.nix-helpers/hs-notmuch.nix {
+                notmuch = pkgs.notmuch;
```
Comment (frasertweedale) [inline default.nix]: Are these two lines (notmuch, talloc) necessary?

File: default.nix [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    };
+  };
+  pkgs = import <nixpkgs> { inherit config; };
+  env = pkgs.haskell.packages.${compiler}.ghcWithPackages (self: [
+    self.purebred
+  ]);
+  in
+    if pkgs.lib.inNixShell
+    then pkgs.lib.overrideDerivation pkgs.haskell.packages.${compiler}.purebred.env (old: {
+         buildInputs = old.buildInputs ++ [
```
Comment (frasertweedale) [inline default.nix]: Can we clean this up?  Is it necessary?

Comment (romanofski) [reply default.nix]: No technical reason. Just took it out to save a few seconds built time. Will remove it.

File: default.nix (hunk 1 above)
Comment (romanofski) [reply default.nix]: No technical reason. Just took it out to save a few seconds built time. Will remove it.

File: default.nix (hunk 3 above)
Comment (romanofski) [reply default.nix]: Ignore the previous comment. Do you mean the commented out line? Yeah I thought I'd keep it as an example, but yeah - happy to remove it too.

File: default.nix (hunk 2 above)
Comment (romanofski) [reply default.nix]: Yes. It specifies which dependency notmuch is coming from (hs-notmuch depends on the C-sources named notmuch). More info I got from: https://github.com/Gabriel439/haskell-nix/blob/9c72b6ecbc5e25df509dfd6ee3d5ee8b9eb21f14/project2/README.md

File: default.nix (hunk 3 above)
Comment (romanofski) [reply default.nix]: Done.

File: default.nix (hunk 1 above)
Comment (romanofski) [reply default.nix]: Done.

**Comment by frasertweedale** [comment]
There are now conflicts with README.  I'm happy for you to merge after resolving.  I think there are a few improvements that can be made but better to merge and I can propose them separately.

**Pull request merged**